### PR TITLE
automatic transform application

### PR DESCRIPTION
**Pull request by jsalvatier**
not ready to merge. 

This copies stan in automatically applying transforms to bounded variables. Anecdotally this seems to make some models work much better with NUTS. For example this one: https://gist.github.com/fonnesbeck/4f0e9d8a5e6b69556233

Also changes the syntax for applying transforms to `x = HalfNormal('x', 0, 1, transform=logtransform)`. Which seems dramatically easier. We should maybe also clean up the transforms.

### CODE REVIEW

File: pymc3/distributions/continuous.py [hunk 1]
```diff
@@ -46,12 +59,15 @@ class Uniform(Continuous):
...
-    def __init__(self, lower=0, upper=1, *args, **kwargs):
+    def __init__(self, lower=0, upper=1, transform=None, *args, **kwargs):
         super(Uniform, self).__init__(*args, **kwargs)
+        
         self.lower = lower
         self.upper = upper
         self.mean = (upper + lower) / 2.
         self.median = self.mean
+        if transform is None:
+            self.transform=interval_transform(lower, upper)
```
Comment (twiecki) [inline pymc3/distributions/continuous.py]: nitpicking: pep8 requires spaces here

File: pymc3/distributions/continuous.py [hunk 2]
```diff
@@ -132,11 +148,12 @@ class HalfNormal(Continuous):
       - `sd` : sd > 0 (alternative parameterization)
 
     """
-    def __init__(self, tau=None, sd=None, *args, **kwargs):
+    def __init__(self, tau=None, sd=None, transform=logtransform, *args, **kwargs):
```
Comment (twiecki) [inline pymc3/distributions/continuous.py]: Isn't the transform already specified in `PositiveContinuous` here? Shouldn't have to specify it twice.

**Comment by twiecki** [comment]
This is amazing.

Does this still add the transformed sampler to the trace? This might be confusing to users and also increases mem usage unnecessarily.

**Comment by fonnesbeck** [comment]
So, if for some reason you don't need/want a log transformation on the sampling space of the HalfNormal (say the values are all really large), you'd just pass `transform=None`? I assume Dirichlet will also have `transform=simplextransform`?

I like it.

File: pymc3/backends/base.py [hunk 3]
```diff
@@ -40,6 +44,9 @@ def __init__(self, name, model=None, vars=None):
                            for var, value in var_values}
         self.chain = None
 
+    def values(self, point):
+        return (point[v] for v in self.varnames)
+
     ## Sampling methods
```
Comment (kyleam) [inline pymc3/backends/base.py:44]: I'd prefer to name this something else.  I think it looks too much like an attribute this way.

File: pymc3/distributions/continuous.py (hunk 1 above)
Comment (jsalvatier) [inline pymc3/distributions/continuous.py]: noted.

File: pymc3/distributions/continuous.py (hunk 2 above)
Comment (jsalvatier) [inline pymc3/distributions/continuous.py]: good catch

File: pymc3/backends/base.py (hunk 3 above)
Comment (jsalvatier) [inline pymc3/backends/base.py:44]: good point. I wasn't quite sure what to call it. I'll think about it a bit more. It should probably have a leading underscore too

**Comment by jsalvatier** [comment]
@fonnesbeck yup, that's exactly right. Good point, I forgot about Dirichlet.

@twiecki, yes it does right now. Good point that it might be confusing. I hadn't even considered the possibility of not recording the inputs at all, just the untransformed variables. That does sound attractive from the user point of view. I would want a way to turn it on for all of them, for if you're debugging the sampling I think it might be helpful. Also, it means you'd have to figure out how to restart sampling from a trace, since the input values aren't stored. I think I'm going to leave it out for now, but I definitely want to figure out how to make this work. If you have ideas, please feel free to have a crack.

**Comment by jsalvatier** [comment]
@fonnesbeck @twiecki, the problem of making the transformations easy to understand for the user seems a bit tricky. 

We could remove the transformed variables from the trace by default, but then there needs to be a way to get a starting value from those. The values returned from `find_MAP` and `find_hessian` are both going to return values from the transformed space, which might be confusing.

I wonder if we can solve this problem by transforming values (transformed space -> original space) any time we return something up to the user. And then reversing the transformation when a user passes in something with values in the original space.

I worry that this won't fail gracefully. That if we don't do this literally everywhere it will be extra confusing when the transformed values leak through. Not sure what to do about that though. 

I've solved the problem of having to pass in the transformed variables to `find_MAP` or `NUTS` by finding the parent inputs for every variable and using those variables. That will also help solve the a similar problem for missing variables.

**Comment by twiecki** [comment]
@jsalvatier Those are valid issues that would be hard to work around. I agree that we should punt on the option of not saving the transformed variables and move ahead without it.

File: pymc3/distributions/continuous.py [hunk 4]
```diff
@@ -46,13 +57,17 @@ class Uniform(Continuous):
     upper : float
         Upper limit (defaults to 1)
     """
-    def __init__(self, lower=0, upper=1, *args, **kwargs):
+    def __init__(self, lower=0, upper=1, transform='interval', *args, **kwargs):
```
Comment (twiecki) [inline pymc3/distributions/continuous.py:60]: Should this not be just `interval` instead of `'interval'`?

Comment (twiecki) [inline pymc3/distributions/continuous.py:60]: I figured it out, it's correct (but a bit odd).

Comment (jsalvatier) [inline pymc3/distributions/continuous.py:60]: any suggestions?

**Comment by jsalvatier** [comment]
Can someone take a good look at this?

File: pymc3/examples/dirichlet.py [hunk 5]
```diff
@@ -7,9 +7,7 @@
     k = 5
     a = constant(np.array([2, 3., 4, 2, 2]))
 
-    p, p_m1 = model.TransformedVar(
-        'p', Dirichlet.dist(a, shape=k),
-        simplextransform)
+    p = Dirichlet('p', a, shape=k)
```
Comment (fonnesbeck) [inline pymc3/examples/dirichlet.py:10]: I love this. So much better.

**Comment by twiecki** [comment]
@jsalvatier I took another look and I really like the implementation. Very clear so from my end good to merge!

File: pymc3/backends/base.py [hunk 6]
```diff
@@ -31,6 +31,7 @@ def __init__(self, name, model=None, vars=None):
         self.varnames = [str(var) for var in vars]
         self.fn = model.fastfn(vars)
 
+
         ## Get variable shapes. Most backends will need this
```
Comment (kyleam) [inline pymc3/backends/base.py:35]: Daring, minimalistic, and profound at the same time.

Comment (jsalvatier) [inline pymc3/backends/base.py:35]: lol, oops.

**Comment by twiecki** [comment]
:+1:, this is an amazing feature.

**Pull request merged**